saml2/saml2-service-provider/src/opensaml4Main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java

```diff
@@ -359,12 +359,19 @@ public void setResponseAuthenticationConverter(
 			Response response = responseToken.getResponse();
 			Saml2AuthenticationToken token = responseToken.getToken();
 			Saml2ResponseValidatorResult result = Saml2ResponseValidatorResult.success();
+			AbstractSaml2AuthenticationRequest request = token.getAuthenticationRequest();
 			String statusCode = getStatusCode(response);
 			if (!StatusCode.SUCCESS.equals(statusCode)) {
 				String message = String.format("Invalid status [%s] for SAML response [%s]", statusCode,
 						response.getID());
 				result = result.concat(new Saml2Error(Saml2ErrorCodes.INVALID_RESPONSE, message));
 			}
+			String inResponseTo = response.getInResponseTo();
+			if (StringUtils.hasText(inResponseTo) && request != null && !inResponseTo.equals(request.getId())) {
```

Comment:
Typically in Spring Security, if an if statement has 3+ clauses, we break it up into a separate method, for example (pseudocode):

```java
private static Saml2ResponseValidationResult validateInResponseTo(Response response, AbstractAuthenticationRequest request) {
    String inReponseTo = response.getInResponseTo();
    if (!StringUtils.hasText(inResponseTo)) {
        return Saml2ResponseValidationResult.success();
    }
    if (request == null) {
        return Saml2ResponseValidationResult.success();
    }
    // ....
    return Saml2ResponseValidatorResult.error(new Saml2Error(...));
}
```